src/fidesops/migrations/versions/7abe778b7082_update_fideslang_data_categories.py

```diff
@@ -0,0 +1,217 @@
+    )
+
+    return privacy_key in removed
+
+
+def is_removed_upgrade(privacy_key: str) -> bool:
+    removed = (
+        "account",
+        "account.contact",
+        "account.contact.city",
+        "account.contact.country",
+        "account.contact.email",
+        "account.contact.phone_number",
+        "account.contact.postal_code",
+        "account.contact.state",
+        "account.contact.street",
+        "account.payment.financial_account_number",
+        "user.derived",
+        "user.derived.identifiable",
+        "user.derived.nonidentifiable",
+        "user.derived.nonidentifiable.sensor",
+        "user.provided",
+        "user.provided.identifiable",
+        "user.provided.nonidentifiable",
+    )
+
+    return privacy_key in removed
+
+
+def update_field(field: dict, up_down: str) -> dict:
+    data_categories = field.get("data_categories")
+    updated = []
+    if up_down == "up":
+        if data_categories:
+            for data_category in data_categories:
+                if is_removed_upgrade(data_category):
+                    logger.info(
+                        "Removing %s, this is no longer a valid category", data_category
+                    )
+                if data_category:
+                    if not is_removed_upgrade(data_category):
+                        new = DATA_MAP_UPGRADE.get(data_category)
+                        if new:
+                            updated.append(new)
+                        else:
+                            updated.append(data_category)
+                    else:
+                        updated.append(data_category)
+        field["data_categories"] = updated
+
+        if field.get("fields"):
+            for field in field["fields"]:
+                update_field(field, up_down)
+    else:
+        if data_categories:
+            for data_category in data_categories:
+                if is_removed_downgrade(data_category):
+                    logger.info(
+                        "Removing %s, this is no longer a valid category", data_category
+                    )
+                if data_category:
+                    if not is_removed_downgrade(data_category):
+                        new = DATA_MAP_DOWNGRADE.get(data_category)
+                        if new:
+                            updated.append(new)
+                        else:
+                            updated.append(data_category)
+                    else:
+                        updated.append(data_category)
+        field["data_categories"] = updated
+
+        # Document databases like mongo can have nested fields so this is the reason
+        # for the recursion here.
+        if field.get("fields"):
+            for field in field["fields"]:
+                update_field(field, up_down)
+
+    return field
+
+
+def upgrade() -> None:
+    sessionlocal = get_db_session(config)
+    with sessionlocal() as session:
+        try:
+            datasets = DatasetConfig.all(db=session)
```

Comment:
Something I've found really useful with (potentially) large data migrations in the past is some sense of a progress indicator via logging. Now we've fetched the entire dataset it'd be good to log the count here, and current iteration below (or maybe every `x` to avoid spamming the logs).

Reply:
Because of the recursion it could be tricky to give an accurate count. The only length we will know is the top level, but we won't know how deep each level goes. Thoughts on how to handle that?

Reply:
Great point. I don't have a good solution for that. As long as the user gets some input, anything is better than nothing.

Reply:
I added some more logging. How helpful it will be is going to depend on how deep vs how wide things are. See if you think it will be good enough.